tfhe/src/high_level_api/tests/gpu_selection.rs

```diff
@@ -0,0 +1,114 @@
+    let keys = ClientKey::generate(config);
+    let compressed_server_keys = CompressedServerKey::new(&keys);
+
+    let mut rng = rand::thread_rng();
+
+    let last_gpu = GpuIndex(GpuIndex::num_gpus() - 1);
+
+    let clear_a: u32 = rng.gen();
+    let clear_b: u32 = rng.gen();
+
+    let mut a = FheUint32::try_encrypt(clear_a, &keys).unwrap();
+    let mut b = FheUint32::try_encrypt(clear_b, &keys).unwrap();
+
+    assert_eq!(a.current_device(), Device::Cpu);
+    assert_eq!(b.current_device(), Device::Cpu);
+    assert_eq!(a.gpu_indexes(), &[]);
+    assert_eq!(b.gpu_indexes(), &[]);
+
+    let cuda_key = compressed_server_keys.decompress_to_specific_gpu(last_gpu);
+
+    set_server_key(cuda_key);
+    let c = &a + &b;
+    let decrypted: u32 = c.decrypt(&keys);
+    assert_eq!(c.current_device(), Device::CudaGpu);
+    assert_eq!(c.gpu_indexes(), &[last_gpu]);
+    assert_eq!(decrypted, clear_a.wrapping_add(clear_b));
+
+    // Check explicit move, but first make sure input are on Cpu still
+    assert_eq!(a.current_device(), Device::Cpu);
+    assert_eq!(b.current_device(), Device::Cpu);
+    assert_eq!(a.gpu_indexes(), &[]);
+    assert_eq!(b.gpu_indexes(), &[]);
+
+    a.move_to_current_device();
+    b.move_to_current_device();
+
+    assert_eq!(a.current_device(), Device::CudaGpu);
+    assert_eq!(b.current_device(), Device::CudaGpu);
+    assert_eq!(a.gpu_indexes(), &[last_gpu]);
+    assert_eq!(b.gpu_indexes(), &[last_gpu]);
+
+    let c = &a + &b;
+    let decrypted: u32 = c.decrypt(&keys);
+    assert_eq!(c.current_device(), Device::CudaGpu);
+    assert_eq!(c.gpu_indexes(), &[last_gpu]);
+    assert_eq!(decrypted, clear_a.wrapping_add(clear_b));
+}
+
+#[test]
+fn test_gpu_selection_2() {
+    if GpuIndex::num_gpus() < 2 {
+        // This test is only really useful if there are 2 GPUs
+        return;
+    }
+    let config = ConfigBuilder::default().build();
+    let keys = ClientKey::generate(config);
+    let compressed_server_keys = CompressedServerKey::new(&keys);
+
+    let mut rng = rand::thread_rng();
+
+    let first_gpu = GpuIndex(0);
+    let last_gpu = GpuIndex(GpuIndex::num_gpus() - 1);
+
+    let clear_a: u32 = rng.gen();
+    let clear_b: u32 = rng.gen();
+
+    let mut a = FheUint32::try_encrypt(clear_a, &keys).unwrap();
+    let mut b = FheUint32::try_encrypt(clear_b, &keys).unwrap();
+
+    assert_eq!(a.current_device(), Device::Cpu);
+    assert_eq!(b.current_device(), Device::Cpu);
+    assert_eq!(a.gpu_indexes(), &[]);
+    assert_eq!(b.gpu_indexes(), &[]);
+
+    let cuda_key = compressed_server_keys.decompress_to_specific_gpu(last_gpu);
+    set_server_key(cuda_key);
+
+    a.move_to_current_device();
+    b.move_to_current_device();
+
+    assert_eq!(a.current_device(), Device::CudaGpu);
+    assert_eq!(b.current_device(), Device::CudaGpu);
+    assert_eq!(a.gpu_indexes(), &[last_gpu]);
+    assert_eq!(b.gpu_indexes(), &[last_gpu]);
+
+    let c = &a + &b;
```

Comment:
could loop over all GPU indices, also allows removing the early return requiring at least 2 GPUs

Reply:
Hum but the test is mostly useful if there are 2 GPUs, because it does some computation on GpuX, then some other computation on GpuY, then tries to access data that is on GpuX when developping the feature, this would make the app abort (that's why the CudaStreamPool exists)

Reply:
ok maybe I misread the goal of the test